packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -243,11 +243,17 @@ export class RunningSummarizer extends TypedEventEmitter<ISummarizerEvents> impl
 			}
 		});
 
+		const immediatelyRefreshLatestSummaryAck =
+			this.mc.config.getBoolean("Fluid.Summarizer.immediatelyRefreshLatestSummaryAck") ??
+			false;
 		this.generator = new SummaryGenerator(
 			this.pendingAckTimer,
 			this.heuristicData,
 			this.submitSummaryCallback,
-			() => {
+			async (options: IRefreshSummaryAckOptions) => {
+				if (immediatelyRefreshLatestSummaryAck) {
+					await this.refreshLatestSummaryAckCallback(options);
```

Comment:
This fix seems fine but it introduces another checkpoint into the system when we have other patterns to already handle these, i.e., `summarizingLock`.
The `summarizingLock` exists to ensure that summary submission and refreshing an ack are mutually exclusive and they happen in correct orders. I believe the real bug here is that in the `afterSummaryAction` function, there is an attempt to run the summaryHeuristics synchronously holding the same `summarizingLock` that was used to run the summary. That logic should be async to give any other process holding the lock to go through first.

Even with this fix, that bug exists where summary submission can hold the lock indefinitely.

Reply:
I see what you mean. Ok that is really helpful. I'll look into this deeper.

Reply:
I would suggest trying that fix out. I am not 100% sure that will fix everything. Every time I look at the logic around summarizing + refreshing, I always learn something that I did not expect :-).